### PR TITLE
Add autograd support for square (including complex) + Sparse CSR support

### PR DESCRIPTION
**Pull request by krshrimali**
This PR is a follow-up PR from https://github.com/pytorch/pytorch/pull/69292, to add derivative support for `square`. This also adds support automatic differentiation for outputs with complex dtypes, along with `square` now supporting Sparse CSR layout.

- ~I've skipped forward mode AD tests, for now, will have to take a look at how to add that for `square`. Hopefully in another PR.~ Forward mode AD support has been added as well.
- ~Instead of writing `2 * grad * self.conj()`, I thought I'll just use `pow_backward(grad, self, 2)` in the formula. I assume both are the same? (please correct if I'm wrong)~ Used the formula instead, based on suggestion: https://github.com/pytorch/pytorch/pull/69510#issuecomment-987693858.

Note: Once #69292 is merged, this PR will be limited to the changes made for `square`. For CI to run, I've added changes from #69292 in this PR as well. To make the review easy, I've marked the relevant changes by review comments. Thanks!

cc @nikitaved @pearu @cpuhrsch @IvanYashchuk @ezyang @albanD @zou3519 @gqchen @soulitzer @Lezcano @Varal7

### CODE REVIEW

File: tools/autograd/gen_variable_type.py [hunk 1]
```diff
@@ -89,7 +89,7 @@
     'triu', 'chunk', 'zero_', 'eq_', 'ne_', 'add', '__radd__', 'sum',
     '_conj', 'sin', 'cos', 'mul', 'sinc', 'sinh', 'cosh', '__rmul__',
     'sgn', 'asin', 'acos', 'sub', 'div', 'cat', 'view_as_complex',
-    'neg', 'complex', 'select', '_s_where', 'as_strided', 'slice', 'constant_pad_nd',
+    'neg', 'complex', 'select', '_s_where', 'as_strided', 'slice', 'square', 'constant_pad_nd',
```
Comment (krshrimali) [inline tools/autograd/gen_variable_type.py:92]: relevant change

File: torch/testing/_internal/common_methods_invocations.py [hunk 2]
```diff
@@ -11624,6 +11645,9 @@ def ref_pairwise_distance(input1, input2):
                        # Reference: https://github.com/pytorch/pytorch/pull/52551#issuecomment-782596181
                        DecorateInfo(unittest.skip("Skipped!"), 'TestUnaryUfuncs', 'test_reference_numerics_hard',
                                     dtypes=[torch.bfloat16]),
+                       # Forward Mode AD support not added yet for square
+                       DecorateInfo(unittest.skip("Skipped!"), 'TestGradients', 'test_inplace_forward_mode_AD'),
+                       DecorateInfo(unittest.skip("Skipped!"), 'TestGradients', 'test_forward_mode_AD'),
```
Comment (krshrimali) [inline torch/testing/_internal/common_methods_invocations.py:11650]: relevant change

File: tools/autograd/derivatives.yaml [hunk 3]
```diff
@@ -1222,6 +1222,10 @@
   self: -0.5 * grad * result.pow(3).conj()
   result: auto_element_wise
 
+- name: square(Tensor self) -> Tensor
+  self: pow_backward(grad, self, 2)
+  result: auto_element_wise
+
```
Comment (krshrimali) [inline tools/autograd/derivatives.yaml:1228]: relevant change

File: torch/testing/_internal/common_methods_invocations.py [hunk 4]
```diff
@@ -11608,6 +11628,7 @@ def ref_pairwise_distance(input1, input2):
                    dtypes=all_types_and_complex_and(torch.bool, torch.float16, torch.bfloat16),
                    decorators=(precisionOverride({torch.complex64: 3e-4, torch.bfloat16: 3e-1}),),
                    supports_forward_ad=True,
+                   supports_sparse_csr=True,
```
Comment (krshrimali) [inline torch/testing/_internal/common_methods_invocations.py:11631]: relevant change

File: aten/src/ATen/native/sparse/SparseCsrTensorMath.cpp [hunk 5]
```diff
@@ -164,17 +173,70 @@ bool is_square_or_vec(int64_t dim_i, int64_t dim_j, int64_t dim_k) {
...
+CREATE_UNARY_UFUNC(log1p);
+CREATE_UNARY_UFUNC(neg);
+CREATE_UNARY_UFUNC(rad2deg);
+CREATE_UNARY_UFUNC(round);
+CREATE_UNARY_UFUNC(sign);
+CREATE_UNARY_UFUNC(sin);
+CREATE_UNARY_UFUNC(sinh);
+CREATE_UNARY_UFUNC(sgn);
+CREATE_UNARY_UFUNC(sqrt);
+CREATE_UNARY_UFUNC(square);
```
Comment (krshrimali) [inline aten/src/ATen/native/sparse/SparseCsrTensorMath.cpp:225]: relevant change

File: aten/src/ATen/native/native_functions.yaml [hunk 6]
```diff
@@ -4283,18 +4336,26 @@
...
   device_check: NoCheck   # TensorIterator
   variants: function, method
+  dispatch:
+    CPU, CUDA: square_
+    SparseCsrCPU, SparseCsrCUDA: square_sparse_csr_
 
 - func: square.out(Tensor self, *, Tensor(a!) out) -> Tensor(a!)
   dispatch:
     CPU, CUDA: square_out
+    SparseCsrCPU, SparseCsrCUDA: square_sparse_csr_out
```
Comment (krshrimali) [inline aten/src/ATen/native/native_functions.yaml:4358]: relevant change

**Comment by Lezcano** [comment]
Please don't assume the backward for pow is the same as the formula. If the formula is simple, please write it inline, as that way we know that it's as fast as it can be.
You added forward AD support via `the `auto_element_wise` keyword. That should be enough. What's the errors that you get from the tests?

File: torch/testing/_internal/common_methods_invocations.py (hunk 2 above)
Comment (IvanYashchuk) [reply torch/testing/_internal/common_methods_invocations.py]: What is the error here? You've included `result: auto_element_wise` in the `derivatives.yaml` so forward AD should be working.

**Review by IvanYashchuk — COMMENTED** [review]
`torch.square` (`at::square`) is a composite function. There's no need on specializing the dispatcher for this. Unfortunately, it's also a breaking change to use 
```
  dispatch:
    CPU, CUDA: square
```
To get `torch.square` working for sparse csr tensor correctly `torch.pow` needs to be implemented.

**Comment by krshrimali** [comment]
> Please don't assume the backward for pow is the same as the formula. If the formula is simple, please write it inline, as that way we know that it's as fast as it can be.

Sure, makes sense. Should be fixed now.

> You added forward AD support via `the `auto_element_wise` keyword. That should be enough. What's the errors that you get from the tests?

Ah, the failures were before I added `auto_element_wise`. I've removed the skips now. Thanks!

**Comment by IvanYashchuk** [comment]
This PR is closed in favor of https://github.com/pytorch/pytorch/pull/69603.

**Pull request closed without merging**